### PR TITLE
feat: allow contract to contract read

### LINKED ISSUE #453: Contract-to-contract read
Add contract to contract read capabilities on accepted and finalized state.
For reads, the usage should be something like:
```python
from backend.node.genvm.contract import Contract
TOKEN_1_CONTRACT_ADDRESS = "0x..."
TOKEN_2_CONTRACT_ADDRESS = "0x..."
... 
def_get_token_balances(self, account_address: str) -> str:
        token_1_contract = Contract(TOKEN_1_CONTRACT_ADDRESS)
        token_1_balance = token_1_contract.balace_of(account_address)

        token_2_contract = Contract(TOKEN_2_CONTRACT_ADDRESS)
        token_2_balance = token_2_contract.balace_of(account_address)
        
        return {token_1: token_1_balance, token_2: token_2_balance}
```

Internally, when instantiated, the Contract class should fetch the ContractSnapshot of the contract at the given address.
Another possibility is that the node understands that this method calls another contract at the time of executing the transaction and fetches that ContractSnapshot

### PR DESCRIPTION
**Pull request by AgustinRamiroDiaz**
Fixes #453

# What

- allows contract to contract reads

# Why

To allow for more possibilities of projects in GenLayer

# Testing done

Added e2e tests

# Decisions made

# Checks

- [ ] I have tested this code
- [ ] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

# Reviewing tips

Don't check the commits

# User facing release notes

Now you can call get methods from other contracts from within your Intelligent Contracts

### CODE REVIEW

File: .github/workflows/backend_integration_tests_pr.yml [hunk 1]
```diff
@@ -99,7 +99,7 @@ jobs:
           pip install -r requirements.test.txt
 
       - name: Run tests
-        run: pytest tests/integration/ -vv
+        run: pytest tests/integration/ -svv
```
Comment (AgustinRamiroDiaz) [inline .github/workflows/backend_integration_tests_pr.yml:102]: extra: print test `print()`s

File: backend/database_handler/contract_snapshot.py [hunk 2]
```diff
@@ -24,12 +26,17 @@ def __init__(self, contract_address: str, session: Session):
...
             .filter(CurrentState.id == self.contract_address)
-            .one()
+            .one_or_none()
         )
 
+        if result is None:
+            raise Exception(f"Contract {self.contract_address} not found")
+
+        return result
+
```
Comment (AgustinRamiroDiaz) [inline backend/database_handler/contract_snapshot.py:39]: extra: improvement of error

File: backend/node/genvm/base.py [hunk 3]
```diff
@@ -18,22 +20,30 @@
...
             "contract_runner": None,
             "VectorStore": VectorStore,
         }
     )
+    if override_globals:
+        globals().update(override_globals)
     try:
         yield
     finally:
+        globals().clear()
```
Comment (AgustinRamiroDiaz) [inline backend/node/genvm/base.py:36]: extra: fix, we were not clearing them, only expanding

File: backend/node/genvm/base.py [hunk 4]
```diff
@@ -347,40 +374,60 @@ def generate_abi_from_schema_methods(contract_schema_methods: dict) -> list:
...
+    ):
+        self.address = address
+
+        self.contract_snapshot = contract_snapshot_factory(address)
+        self.contract_snapshot_factory = contract_snapshot_factory
+
+    def __getattr__(self, name):
+        def method(*args, **kwargs):
+            if not re.match("get_", name):
+                raise Exception("Method name must start with 'get_'")
```
Comment (AgustinRamiroDiaz) [inline backend/node/genvm/base.py:423]: For now we only allow reads

File: docker/Dockerfile.frontend [hunk 5]
```diff
@@ -2,9 +2,7 @@ FROM node:21.7.3-alpine3.20 AS base
 
 WORKDIR /app
 COPY ./frontend/package*.json .
-RUN npm ci --cache /tmp/empty-cache && \
-    npm cache clean --force && \
-    rm -rf /tmp/empty-cache
+RUN --mount=type=cache,target=/root/.npm npm ci
```
Comment (AgustinRamiroDiaz) [inline docker/Dockerfile.frontend:5]: extra: better caching with buildkit

File: backend/node/base.py [hunk 6]
```diff
@@ -83,7 +89,28 @@ async def run_contract(
...
+        result = GenVM.get_contract_data(
+            code,
+            state,
+            method_name,
+            method_args,
+            self.contract_snapshot_factory,
+            output_buffer,
+        )
+
+        if self.genvm.contract_runner.mode == ExecutionMode.LEADER:
```
Comment (cristiam86) [inline backend/node/base.py:103]: @AgustinRamiroDiaz this will conflict with https://github.com/yeagerai/genlayer-simulator/pull/540

**Review by cristiam86 — APPROVED** [review]
great job, but I needed like 30 mins to understand the tests :D

**Pull request merged**